Check for fused kernel before inplace update

exclamaforte opened the pull request:
Summary:
Given an op, with a pair (output buffer, input buffer) from that op, we consider marking the output buffer as inline. However, if the parent of input buffer and the current op are going to be fused, then we don't want to mark the output buffer as inline. This change checks that criterion, and skips inlining if it is so.

Test Plan:
New unit test "layer_norm_should_not_inplace" runs LayerNorm and checks for no "in_out" pointers.

Fixes #120217

Here's a diagram of the issue:
![Inline+Fusion](https://github.com/user-attachments/assets/c03308d8-fdbf-40a0-a46d-964ece5f9e6d)



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

eellison commented on torch/_inductor/scheduler.py:
```diff
@@ -377,7 +377,7 @@ def can_inplace(self, read_dep: dependencies.Dep) -> bool:
         return False
 
-    def decide_inplace_update(self) -> None:
+    def decide_inplace_update(self, fused_nodes: Optional[Set[str]] = None) -> None:
```
We don't need to pass this through. You can access the fused node through: `self.scheduler.name_to_fused_node`

eellison commented on torch/_inductor/scheduler.py:
```diff
@@ -419,6 +419,11 @@ def decide_inplace_update(self) -> None:
                     and not isinstance(input_buf.defining_op, NopKernelSchedulerNode)
                 ):
+                    # If the writers of input_buf are in the same FusedSchedulerNode as the current op, then there is
+                    # no need to inline.
```
nit: /s/inline/inplace.

eellison reviewed (commented):
looks good, one simplification on function calling convention

exclamaforte commented:
@pytorchbot label "topic: not user facing"

exclamaforte committed:
Check for fused kernel before inplace update

Summary:
Given an op, with a pair (output buffer, input buffer) from that op, we consider marking the output buffer as inline. However, if the parent of input buffer and the current op are going to be fused, then we don't want to mark the output buffer as inline. This change checks that criterion, and skips inlining if it is so.

Test Plan:
New unit test "layer_norm_should_not_inplace" runs LayerNorm and checks for no "in_out" pointers.

Fixes #120217

exclamaforte commented:
@pytorchbot merge
